lib/middlewares/src/metering.rs

```diff
@@ -0,0 +1,158 @@
+            .global_initializers
+            .push(GlobalInit::I64Const(self.initial_limit as i64));
+    }
+}
+
+impl<F: Fn(&Operator) -> u64 + Copy + Clone + Send + Sync> fmt::Debug for FunctionMetering<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FunctionMetering")
+            .field("cost_function", &"<function>")
+            .field("remaining_points_index", &self.remaining_points_index)
+            .finish()
+    }
+}
+
+impl<F: Fn(&Operator) -> u64 + Copy + Clone + Send + Sync> FunctionMiddleware
+    for FunctionMetering<F>
+{
+    fn feed<'a>(
+        &mut self,
+        operator: Operator<'a>,
+        state: &mut MiddlewareReaderState<'a>,
+    ) -> WpResult<()> {
+        // Get the cost of the current operator, and add it to the accumulator.
+        // This needs to be done before the metering logic, to prevent operators like `Call` from escaping metering in some
+        // corner cases.
+        self.accumulated_cost += (self.cost_function)(&operator);
+
+        // Possible sources and targets of a branch. Finalize the cost of the previous basic block and perform necessary checks.
+        match operator {
+            Operator::Loop { .. } // loop headers are branch targets
+            | Operator::End // block ends are branch targets
+            | Operator::Else // "else" is the "end" of an if branch
+            | Operator::Br { .. } // branch source
+            | Operator::BrTable { .. } // branch source
+            | Operator::BrIf { .. } // branch source
+            | Operator::Call { .. } // function call - branch source
+            | Operator::CallIndirect { .. } // function call - branch source
+            | Operator::Return // end of function - branch source
+            => {
+                if self.accumulated_cost > 0 {
+                    // if unsigned(globals[remaining_points_index]) < unsigned(self.accumulated_cost) { throw(); }
+                    state.push_operator(Operator::GlobalGet { global_index: self.remaining_points_index.as_u32() });
+                    state.push_operator(Operator::I64Const { value: self.accumulated_cost as i64 });
+                    state.push_operator(Operator::I64LtU);
+                    state.push_operator(Operator::If { ty: WpTypeOrFuncType::Type(WpType::EmptyBlockType) });
+                    state.push_operator(Operator::Unreachable); // FIXME: Signal the error properly.
+                    state.push_operator(Operator::End);
+
+                    // globals[remaining_points_index] -= self.accumulated_cost;
+                    state.push_operator(Operator::GlobalGet { global_index: self.remaining_points_index.as_u32() });
+                    state.push_operator(Operator::I64Const { value: self.accumulated_cost as i64 });
+                    state.push_operator(Operator::I64Sub);
+                    state.push_operator(Operator::GlobalSet { global_index: self.remaining_points_index.as_u32() });
```

Comment:
Might be ergonomic to implement https://doc.rust-lang.org/std/iter/trait.Extend.html for this, so you can put a bunch of operators in a collection like `&[]` and push them all in

Reply:
Done. Two new `Extend`s.